File: packages/soulseek_protocol/lib/src/services/room_chat_service.dart
import 'dart:async';

import '../connection/server_connection.dart';
import '../messages/buffer.dart';
import '../messages/codes.dart';
import '../messages/message.dart';

class RoomChatService {
  final ServerTransport _server;
  final _roomMessagesController = StreamController<RoomMessage>.broadcast();
  final _userJoinedController = StreamController<UserJoinedRoom>.broadcast();
  final _userLeftController = StreamController<UserLeftRoom>.broadcast();
  final _roomListController = StreamController<RoomList>.broadcast();

  late final StreamSubscription _messageSub;

  RoomChatService({required ServerTransport server}) : _server = server;

  Stream<RoomMessage> get roomMessages => _roomMessagesController.stream;
  Stream<UserJoinedRoom> get userJoined => _userJoinedController.stream;
  Stream<UserLeftRoom> get userLeft => _userLeftController.stream;
  Stream<RoomList> get roomList => _roomListController.stream;

  void init() {
    _messageSub = _server.messages.listen(_onMessage);
  }

  void joinRoom(String roomName) {
    _server.sendMessage(SoulseekMessage(
      ServerCode.joinRoom,
      JoinRoom(roomName).serialize().toBytes(),
    ));
  }

  void leaveRoom(String roomName) {
    _server.sendMessage(SoulseekMessage(
      ServerCode.leaveRoom,
      LeaveRoom(roomName).serialize().toBytes(),
    ));
  }

  void sendMessage(String roomName, String message) {
    _server.sendMessage(SoulseekMessage(
      ServerCode.roomMessage,
      SendRoomMessage(roomName: roomName, message: message).serialize().toBytes(),
    ));
  }

  void requestRoomList() {
    _server.sendMessage(SoulseekMessage(
      ServerCode.roomList,
      WriteBuffer().toBytes(),
    ));
  }

  void setRoomTicker(String roomName, String ticker) {
    _server.sendMessage(SoulseekMessage(
      ServerCode.roomTickerSet,
      RoomTickerSet(roomName: roomName, ticker: ticker).serialize().toBytes(),
    ));
  }

  void removeRoomTicker(String roomName) {
    _server.sendMessage(SoulseekMessage(
      ServerCode.roomTickerRemove,
      RoomTickerRemove(roomName).serialize().toBytes(),
    ));
  }

  void _onMessage(SoulseekMessage message) {
    try {
      switch (message.code) {
        case ServerCode.roomMessage:
          final msg = RoomMessage.parse(ReadBuffer(message.payload));
          _roomMessagesController.add(msg);
          break;
        case ServerCode.userJoinedRoom:
          final joined = UserJoinedRoom.parse(ReadBuffer(message.payload));
          _userJoinedController.add(joined);
          break;
        case ServerCode.userLeftRoom:
          final left = UserLeftRoom.parse(ReadBuffer(message.payload));
          _userLeftController.add(left);
          break;
        case ServerCode.roomList:
          final list = RoomList.parse(ReadBuffer(message.payload));
          _roomListController.add(list);
          break;
      }
    } catch (_) {
      // Skip malformed room messages
    }
  }

  void dispose() {
    _messageSub.cancel();
    _roomMessagesController.close();
    _userJoinedController.close();
    _userLeftController.close();
    _roomListController.close();
  }
}

File: Soulseek.Protocol/Services/RoomChatService.cs
using Soulseek.Protocol.Connection;
using Soulseek.Protocol.Messages;

namespace Soulseek.Protocol.Services;

public class RoomChatService
{
    private readonly IServerTransport _server;
    private readonly Subject<RoomMessageData> _roomMessagesSubject = new();
    private readonly Subject<UserJoinedRoom> _userJoinedSubject = new();
    private readonly Subject<UserLeftRoom> _userLeftSubject = new();
    private readonly Subject<RoomList> _roomListSubject = new();
    private IDisposable? _messageSub;

    public RoomChatService(IServerTransport server)
    {
        _server = server;
    }

    public IObservable<RoomMessageData> RoomMessages => _roomMessagesSubject;
    public IObservable<UserJoinedRoom> UserJoined => _userJoinedSubject;
    public IObservable<UserLeftRoom> UserLeft => _userLeftSubject;
    public IObservable<RoomList> RoomList => _roomListSubject;

    public void Init()
    {
        _messageSub = _server.Messages.Subscribe(OnMessage);
    }

    public void JoinRoom(string roomName)
    {
        _server.SendMessage(new SoulseekMessage(
            ServerCode.JoinRoom,
            new JoinRoom(roomName).Serialize().ToBytes()
        ));
    }

    public void LeaveRoom(string roomName)
    {
        _server.SendMessage(new SoulseekMessage(
            ServerCode.LeaveRoom,
            new LeaveRoom(roomName).Serialize().ToBytes()
        ));
    }

    public void SendMessage(string roomName, string message)
    {
        _server.SendMessage(new SoulseekMessage(
            ServerCode.RoomMessage,
            new SendRoomMessage(roomName, message).Serialize().ToBytes()
        ));
    }

    public void RequestRoomList()
    {
        _server.SendMessage(new SoulseekMessage(
            ServerCode.RoomList,
            new WriteBuffer().ToBytes()
        ));
    }

    public void SetRoomTicker(string roomName, string ticker)
    {
        _server.SendMessage(new SoulseekMessage(
            ServerCode.RoomTickerSet,
            new RoomTickerSet(roomName, ticker).Serialize().ToBytes()
        ));
    }

    public void RemoveRoomTicker(string roomName)
    {
        _server.SendMessage(new SoulseekMessage(
            ServerCode.RoomTickerRemove,
            new RoomTickerRemove(roomName).Serialize().ToBytes()
        ));
    }

    private void OnMessage(SoulseekMessage message)
    {
        try
        {
            switch (message.Code)
            {
                case ServerCode.RoomMessage:
                    var msg = RoomMessageData.Parse(new ReadBuffer(message.Payload));
                    _roomMessagesSubject.OnNext(msg);
                    break;
                case ServerCode.UserJoinedRoom:
                    var joined = UserJoinedRoom.Parse(new ReadBuffer(message.Payload));
                    _userJoinedSubject.OnNext(joined);
                    break;
                case ServerCode.UserLeftRoom:
                    var left = UserLeftRoom.Parse(new ReadBuffer(message.Payload));
                    _userLeftSubject.OnNext(left);
                    break;
                case ServerCode.RoomList:
                    var list = RoomList.Parse(new ReadBuffer(message.Payload));
                    _roomListSubject.OnNext(list);
                    break;
            }
        }
        catch
        {
            // Skip malformed room messages
        }
    }

    public void Dispose()
    {
        _messageSub?.Dispose();
    }
}